core/src/main/scala/org/bitcoins/core/wallet/builder/RawTxFinalizer.scala

```diff
@@ -462,99 +479,85 @@ object SubtractFeeFromOutputsFinalizer {
-      Vector.fill(txBuilderResult.inputs.length)(
-        P2WPKHWitnessV0(ECPublicKey.freshPublicKey, DummyECDigitalSignature)))
-
-    val wtx = txBuilderResult.toWitnessTransaction(dummyTxWit)
-
-    val outputsAfterFee =
-      SubtractFeeFromOutputsFinalizer.subtractFees(wtx, feeRate, spks)
-
-    Future.successful(
-      txBuilderResult.toBaseTransaction.copy(outputs = outputsAfterFee))
-  }
-}
+case class DualFundingInput(
+    scriptSignature: ScriptSignature,
+    maxWitnessLen: Int)
 
-/** Finalizes a dual-funded transaction given the InputInfos
-  * from both parties and the non-change output ScriptPubKey.
+/** Finalizes a dual-funded transaction given the DualFundingInputs
+  * from both parties, their change spks and the non-change spk.
   *
   * This includes adding the future fee of spending transactions
   * to the non-change output as well as subtracting relevant fees
   * from the change outputs. This finalizer filters dust outputs.
-  *
-  * Note: spendingVBytes may be 0
   */
 case class DualFundingTxFinalizer(
-    spendingVBytes: Long,
-    feeRate: FeeUnit,
-    fundingSPK: ScriptPubKey,
-    inputInfos: Vector[InputInfo])
+    offerInputs: Vector[DualFundingInput],
+    offerPayoutSPK: ScriptPubKey,
+    offerChangeSPK: ScriptPubKey,
+    acceptInputs: Vector[DualFundingInput],
+    acceptPayoutSPK: ScriptPubKey,
+    acceptChangeSPK: ScriptPubKey,
+    feeRate: SatoshisPerVirtualByte,
+    fundingSPK: ScriptPubKey)
     extends RawTxFinalizer {
+  private val sharedFutureFee = (feeRate * 125).satoshis.toLong
 
-  override def buildTx(txBuilderResult: RawTxBuilderResult)(implicit
-      ec: ExecutionContext): Future[Transaction] = {
-    val changeSPKs =
-      txBuilderResult.outputs.map(_.scriptPubKey).filterNot(_ == fundingSPK)
-
-    val addFutureFee = AddFutureFeeFinalizer(
-      fundingSPK,
-      Satoshis(feeRate.toLong * spendingVBytes),
-      changeSPKs)
-    val subtractFee =
-      SubtractFeeFromOutputsFinalizer(inputInfos, feeRate, changeSPKs)
-    val filterDust = FilterDustFinalizer
-
-    addFutureFee
-      .andThen(subtractFee)
-      .andThen(filterDust)
-      .buildTx(txBuilderResult)
+  private val commonFutureFee = if (sharedFutureFee % 2 == 0) {
+    Satoshis(sharedFutureFee / 2)
+  } else {
+    Satoshis(sharedFutureFee / 2 + 1)
   }
-}
 
-/** Finalizes a dual-funded transaction given the non-change output
-  * ScriptPubKey and under the assumption that all funding inputs
-  * from both parties are P2WPKH UTXOs.
-  *
-  * This includes adding the future fee of spending transactions
-  * to the non-change output as well as subtracting relevant fees
-  * from the change outputs. This finalizer filters dust outputs.
-  *
-  * Note: spendingVBytes may be 0
-  */
-case class P2WPKHDualFundingTxFinalizer(
-    spendingVBytes: Long,
-    feeRate: FeeUnit,
-    fundingSPK: ScriptPubKey)
-    extends RawTxFinalizer {
+  private val sharedFundingFee = (feeRate * 53).satoshis.toLong
```

Comment:
can this `53` be a constant, it is not clear what it is for

Reply:
ditto